pynbody/sph/smooth.cpp

```diff
@@ -654,18 +689,20 @@ void smDensity(SMX smx,int pi,int nSmooth,int *pList,float *fList)
 	for (j=0;j<nSmooth;++j) {
 		pj = pList[j];
 		r2 = fList[j]*ih2;
-		rs = 2.0 - sqrt(r2);
-		if (r2 < 1.0) rs = (1.0 - 0.75*rs*r2);
-		else rs = 0.25*rs*rs*rs;
-		if(rs<0) rs=0;
+		if (Wendland) {
```

Comment:
Does this branch inside a critical part of the code have performance implications? If it's only a small percentage, that's fine, but if it makes a big difference to smoothing runtimes I'd prefer to find a different way (e.g. using templates to generate independent code for the different kernels). Have you checked any runtimes?

Reply:
Hi Andrew,
we haven't checked for any runtime performance. This if statement was simply the most straight forward way of checking at runtime which kernel is requested. we are happy to rewrite this using templates but I am not entirely sure what's the best strategy here. shall we define a template for smDensity or a general template for Kernel depending on what's specified by a parameter?

Reply:
It may be that it is having little impact. In the first instance, I'd recommend just measuring the runtime for calculating the density across a large simulation before and after this patch. If that doesn't show a big change, no need for further action.

If it does show a big change, we want the compiler to optimise out the branch, and I can make more specific suggestions if they would be helpful.

Reply:
so, testing on a simulation with about ~1.75e7 particles does not show a timing difference (2 runs: 44s to 43s and 38s to 36s). I guess we can proceed with the code as is for now. However, we realised that changing the config parameter during an interactive session will not trigger recalculation see e.g. density [in sph.__init__.py](https://github.com/pynbody/pynbody/blob/aa4bd2a0e9fb411d68c246daac7a2752e0020d2f/pynbody/sph/__init__.py#L159). I wonder if we should at least add a check somewhere or do something like for the [smoothing length](https://github.com/pynbody/pynbody/blob/aa4bd2a0e9fb411d68c246daac7a2752e0020d2f/pynbody/sph/__init__.py#L146). Not sure what the best practice here would be.